### DISCUSSION TITLE
Difference between Glorys download methods

### DISCUSSION THREAD

**Discussion by uwagura** [NOAA-CEFI-Regional-Ocean-Modeling/seasonal-workflow#4] [General]
It looks like Mercator Ocean provides two different tools for downloading GLORYS data. There is the [FTP Service](https://data.marine.copernicus.eu/product/GLOBAL_MULTIYEAR_PHY_001_030/files), which provides what Mercator Ocean describes as "original files delivered by producers" in the form of daily or monthly netcdf files containing all the variables for a given product, and the [Subsetter Service](https://data.marine.copernicus.eu/product/GLOBAL_MULTIYEAR_PHY_001_030/download?dataset=cmems_mod_glo_phy_myint_0.083deg_P1M-m_202311), which let's you subset files stored as Zarr files in a cloud backed for specific variables,geographic ranges and times.  I assumed these tools delivered the same data, but it looks like the files downloaded via the subset tool are significantly larger than the files downloaded from the FTP service. 

When I download the a days worth of the  **Analysis** product for a single variable  using subset tool, I get files that are usually about ~1.76 GB big. However, when I download the same files using the FTP service, they are usually about ~843 MB big. This is _kinda_ close to half the size, so I initially thought this could be a `float32` vs `float64` issue, but xarray claims that both datasets are stored as `float32`s ( though I believe xarray infers a lot of things under the hood, so maybe xarray is wrong?) 

When I download a days worth of **Reanalysis** data for a single variable using the subset tool, I get files that are usually about ~843 MB big, though the link above claims they should be about 1.76 GB big as well. The FTP files are about 1/3 of the size at ~260 MB  per (3D) variable, but if you click on the link above it will say 1.35 GB since it groups all available variables together. It's unclear to me what could cause this issue since the ratio of sizes isn't a power of 2. 

Is anyone aware of any differences between the data provided by the two services that Mercator Ocean provides, and if we should prefer one over the other? I should note that Kris has been downloading the "original" files from the FTP service, and these are the ones that are currently on the UDA. If no one is away of any difference betweent the services, I may just go ahead and download the FTP data for consistencies sake.

**Comment by yichengt900** [comment]
@uwagura, thank you for the report. I compared the NetCDF (.nc) files downloaded through the two methods you referenced, and here are the primary differences:
```
>nccmp -dmgfs glo12_rg_1d-m_20241001-20241001_3D-thetao_hcst_R20241009.nc cmems_mod_glo_phy-thetao_anfc_0.083deg_P1D-m_thetao_180.00W-179.92E_80.00S-90.00N_0.49-5727.92m_2024-10-01.nc
DIFFER : FILE FORMATS : NC_FORMAT_NETCDF4_CLASSIC <> NC_FORMAT_NETCDF4
DIFFER : NUMBER OF GLOBAL ATTRIBUTES : 10 <> 11
DIFFER : NAME OF GLOBAL ATTRIBUTE : copernicusmarine_version : GLOBAL ATTRIBUTE DOESN'T EXIST IN glo12_rg_1d-m_20241001-20241001_3D-thetao_hcst_R20241009.nc
DIFFER : NAMES OF RECORDS : time <> 
DIFFER : LENGTHS OF RECORDS : time (1) <>  (0)
DIFFER : VARIABLE : time : DIMENSION time IS RECORD IN FILE "glo12_rg_1d-m_20241001-20241001_3D-thetao_hcst_R20241009.nc" BUT NOT IN "cmems_mod_glo_phy-thetao_anfc_0.083deg_P1D-
m_thetao_180.00W-179.92E_80.00S-90.00N_0.49-5727.92m_2024-10-01.nc"
DIFFER : VARIABLE : thetao : DIMENSION time IS RECORD IN FILE "glo12_rg_1d-m_20241001-20241001_3D-thetao_hcst_R20241009.nc" BUT NOT IN "cmems_mod_glo_phy-thetao_anfc_0.083deg_P1
D-m_thetao_180.00W-179.92E_80.00S-90.00N_0.49-5727.92m_2024-10-01.nc"
DIFFER : TYPES : ATTRIBUTE : step : VARIABLE : longitude : FLOAT <> DOUBLE
DIFFER : TYPES : ATTRIBUTE : step : VARIABLE : latitude : FLOAT <> DOUBLE
DIFFER : VARIABLE "time" IS MISSING ATTRIBUTE WITH NAME "long_name" IN FILE "cmems_mod_glo_phy-thetao_anfc_0.083deg_P1D-m_thetao_180.00W-179.92E_80.00S-90.00N_0.49-5727.92m_2024
-10-01.nc"
DIFFER : VARIABLE "time" IS MISSING ATTRIBUTE WITH NAME "standard_name" IN FILE "cmems_mod_glo_phy-thetao_anfc_0.083deg_P1D-m_thetao_180.00W-179.92E_80.00S-90.00N_0.49-5727.92m_
2024-10-01.nc"
DIFFER : LENGTHS : ATTRIBUTE : units : VARIABLE : time : 31 <> 22 : VALUES : "hours since 1950-01-01 00:00:00" : "hours since 1950-01-01"
DIFFER : VARIABLE "time" IS MISSING ATTRIBUTE WITH NAME "axis" IN FILE "cmems_mod_glo_phy-thetao_anfc_0.083deg_P1D-m_thetao_180.00W-179.92E_80.00S-90.00N_0.49-5727.92m_2024-10-0
1.nc"
DIFFER : VARIABLE "time" IS MISSING ATTRIBUTE WITH NAME "valid_min" IN FILE "glo12_rg_1d-m_20241001-20241001_3D-thetao_hcst_R20241009.nc"
DIFFER : VARIABLE "time" IS MISSING ATTRIBUTE WITH NAME "valid_max" IN FILE "glo12_rg_1d-m_20241001-20241001_3D-thetao_hcst_R20241009.nc"
DIFFER : NUMBER OF ATTRIBUTES : VARIABLE : time : 5 <> 4
DIFFER : TYPES : ATTRIBUTE : valid_min : VARIABLE : thetao : FLOAT <> DOUBLE
DIFFER : TYPES : ATTRIBUTE : valid_max : VARIABLE : thetao : FLOAT <> DOUBLE
DIFFER : VARIABLE : time : POSITION : [0] : VALUES : 655260 <> 655248
```
The core data itself (e.g., `thetao`) appears to be consistent. For consistency, I suggest using the FTP data.

**Comment by andrew-c-ross** [comment]
In the data I got through `copernicusmarine subset`, the reanalysis and interim reanalysis are packed, while the analysis/forecast is not:
```
ncdump -h cmems_mod_glo_phy_my_0.083deg_P1D-m_thetao_180.00W-179.92E_80.00S-90.00N_0.49-5727.92m_2020-09-24.nc

dimensions:
        depth = 50 ;
        latitude = 2041 ;
        longitude = 4320 ;
        time = 1 ;
variables:
        short thetao(time, depth, latitude, longitude) ;
                thetao:_FillValue = -32767s ;
                thetao:cell_methods = "area: mean" ;
                thetao:long_name = "Temperature" ;
                thetao:standard_name = "sea_water_potential_temperature" ;
                thetao:unit_long = "Degrees Celsius" ;
                thetao:units = "degrees_C" ;
                thetao:valid_max = 21306LL ;
                thetao:valid_min = -32766LL ;
                thetao:add_offset = 21. ;
                thetao:scale_factor = 0.000732444226741791 ;
```

```
ncdump -h cmems_mod_glo_phy-thetao_anfc_0.083deg_P1D-m_thetao_180.00W-179.83E_80.00S-90.00N_0.49-5727.92m_2024-09-24.nc

dimensions:
        depth = 50 ;
        latitude = 2041 ;
        longitude = 4319 ;
        time = 1 ;
variables:
        float thetao(time, depth, latitude, longitude) ;
                thetao:_FillValue = 9.96921e+36f ;
                thetao:cell_methods = "area: mean" ;
                thetao:long_name = "Temperature" ;
                thetao:standard_name = "sea_water_potential_temperature" ;
                thetao:unit_long = "Degrees Celsius" ;
                thetao:units = "degrees_C" ;
                thetao:valid_max = 40. ;
                thetao:valid_min = -10. ;
```

Unpacked, the reanalysis file balloons in size (this converts it to double, which should be 4x the bytes of short):
```
ncpdq -U cmems_mod_glo_phy_my_0.083deg_P1D-m_thetao_180.00W-179.92E_80.00S-90.00N_0.49-5727.92m_2020-09-24.nc -O unpacked.nc
du -sh unpacked.nc 
3.3G    unpacked.nc
```

In Yi-Cheng's output one of the files is netcdf4, and this may be the secret since if I take the unpacked file and convert it to netcdf4 with compression:
```
ncks -4 -L 5 unpacked.nc -O nc4.nc
du -sh nc4.nc 
744M    nc4.nc
```

**Reply by uwagura** [reply]
Thanks for looking into this. Yeah, I noticed that the files I downloaded via the `subset` tool were in the `netcdf4` format, while the files I downloaded via the `get` tool were in the `netcdf4-classic` format. According to the [ucar users' guide](https://docs.unidata.ucar.edu/nug/current/netcdf_utilities_guide.html#guide_nccopy), files can end up in the `netcdf4-classic` format if you try to compress netcdf files that in the `classic` format, so perhaps that's what happened here? 

In any case, I have forwarded this issue to Copernicus marine, and they said they are looking into it. In the meantime, my plan is to just downloaded files use the FTP tool (i.e using `copernicusmarine get`), since it looks like the data is the same and these files are generally smaller.
